src/functional/otf/stages.py

```diff
@@ -0,0 +1,116 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import dataclasses
+from typing import Any, Generic, Optional, Protocol, TypeVar
+
+from functional.iterator import ir as itir
+from functional.otf import languages
+from functional.otf.source import source
+
+
+SrcL = TypeVar("SrcL", bound=languages.LanguageTag)
+TgtL = TypeVar("TgtL", bound=languages.LanguageTag)
+SettingT = TypeVar("SettingT", bound=languages.LanguageSettings)
+SrcL_co = TypeVar("SrcL_co", bound=languages.LanguageTag, covariant=True)
+TgtL_co = TypeVar("TgtL_co", bound=languages.LanguageTag, covariant=True)
+SettingT_co = TypeVar("SettingT_co", bound=languages.LanguageSettings, covariant=True)
+
+
+@dataclasses.dataclass(frozen=True)
+class ProgramCall:
```

Comment:
I'm wondering if closure is not a more appropriate name. I just brought this up though, I'm happy with `ProgramCall` too.
```suggestion
class ProgramClosure:
```

Reply:
I was wondering the same but was persuaded by https://en.wikipedia.org/wiki/Closure_(computer_programming) that closure is not the correct name. According to that a closure "Stores a function with an *environmnent*" and "environment" is defined as a mapping of values that were in scope at function definition to "free variables" inside the function. That explicitly excludes function parameters as far as I can tell.